src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1309 @@
+//
+// Parameters:
+//   vec    - The vector to binary search in
+//   offset - The offset to search for
+//
+// Returns:
+//    Index of the first entry with an equal offset, or bitwise complement of
+//    first entry with a higher offset.
+//
+template <typename T, unsigned(T::*field)>
+static size_t BinarySearch(const jitstd::vector<T>& vec, unsigned offset)
+{
+    size_t min = 0;
+    size_t max = vec.size();
+    while (min < max)
+    {
+        size_t mid = min + (max - min) / 2;
+        if (vec[mid].*field == offset)
+        {
+            while (mid > 0 && vec[mid - 1].*field == offset)
+            {
+                mid--;
+            }
+
+            return mid;
+        }
+        if (vec[mid].*field < offset)
+        {
+            min = mid + 1;
+        }
+        else
+        {
+            max = mid;
+        }
+    }
+
+    return ~min;
+}
+
+// Represents a single replacement of a (field) access into a struct local.
+struct Replacement
+{
+    unsigned  Offset;
+    var_types AccessType;
+    unsigned  LclNum;
+    // Is the replacement local (given by LclNum) fresher than the value in the struct local?
+    bool NeedsWriteBack = true;
+    // Is the value in the struct local fresher than the replacement local?
+    bool NeedsReadBack = false;
+
+    Replacement(unsigned offset, var_types accessType, unsigned lclNum)
+        : Offset(offset), AccessType(accessType), LclNum(lclNum)
+    {
+    }
+
+    bool Overlaps(unsigned otherStart, unsigned otherSize) const
+    {
+        unsigned end = Offset + genTypeSize(AccessType);
+        if (end <= otherStart)
+        {
+            return false;
+        }
+
+        unsigned otherEnd = otherStart + otherSize;
+        if (otherEnd <= Offset)
+        {
+            return false;
+        }
+
+        return true;
+    }
+};
+
+enum class AccessKindFlags : uint32_t
+{
+    None                    = 0,
+    IsCallArg               = 1,
+    IsAssignmentSource      = 2,
+    IsAssignmentDestination = 4,
+    IsCallArgByImplicitRef  = 8,
+    IsCallArgOnStack        = 16,
+    IsCallRetBuf            = 32,
+    IsReturned              = 64,
+};
+
+inline constexpr AccessKindFlags operator~(AccessKindFlags a)
+{
+    return (AccessKindFlags)(~(uint32_t)a);
+}
+
+inline constexpr AccessKindFlags operator|(AccessKindFlags a, AccessKindFlags b)
+{
+    return (AccessKindFlags)((uint32_t)a | (uint32_t)b);
+}
+
+inline constexpr AccessKindFlags operator&(AccessKindFlags a, AccessKindFlags b)
+{
+    return (AccessKindFlags)((uint32_t)a & (uint32_t)b);
+}
+
+inline AccessKindFlags& operator|=(AccessKindFlags& a, AccessKindFlags b)
+{
+    return a = (AccessKindFlags)((uint32_t)a | (uint32_t)b);
+}
+
+inline AccessKindFlags& operator&=(AccessKindFlags& a, AccessKindFlags b)
+{
+    return a = (AccessKindFlags)((uint32_t)a & (uint32_t)b);
+}
+
+// Tracks all the accesses into one particular struct local.
+class LocalUses
+{
+    jitstd::vector<Access> m_accesses;
```

Comment:
In the long run a sorted vector might not be the right way to go -- it can certainly have bad perf characteristics in pathological cases (e.g. think `InlineArray` with code accessing thousands of elements in reverse order). Maybe a splay tree or any balanced tree would be better. I think for now this suffices, though.